pyro/infer/autoguide/guides.py

```diff
@@ -347,6 +353,157 @@ def median(self, *args, **kwargs):
         return self(*args, **kwargs)
 
 
+class AutoNormal(AutoGuide):
+    """This implementation of :class:`AutoGuide` uses Normal(0, 1) distributions
+    to construct a guide over the entire latent space. The guide does not
+    depend on the model's ``*args, **kwargs``.
+
+    It should be equivalent to :class: `AutoDiagonalNormal` , but with
+    more convenient site names and with better support for
+    :class:`~pyro.infer.trace_mean_field_elbo.TraceMeanField_ELBO` .
+
+    In :class:`AutoDiagonalNormal` , if your model has N named
+    parameters with dimensions k_i and sum k_i = D, you get a single
+    vector of length D for your mean, and a single vector of length D
+    for sigmas.  This guide gives you N distinct normals that you can
+    call by name.
+
+    Usage::
+
+        guide = AutoNormal(model)
+        svi = SVI(model, guide, ...)
+    """
+    def __init__(self, model, init_loc_fn=init_to_feasible, init_scale=0.1):
+        if init_loc_fn is not init_to_feasible:
+            raise NotImplementedError("TODO")
+        self.init_loc_fn = init_loc_fn
+
+        if not isinstance(init_scale, float) or not (init_scale > 0):
+            raise ValueError("Expected init_scale > 0. but got {}".format(init_scale))
+        self._init_scale = init_scale
+
+        model = InitMessenger(self.init_loc_fn)(model)
+        super().__init__(model)
+
+    def _setup_prototype(self, *args, **kwargs):
+        super()._setup_prototype(*args, **kwargs)
+
+        self._unconstrained_shapes = {}
+        self._cond_indep_stacks = {}
+        self.locs = PyroModule(name='locs')
+        self.scales = PyroModule(name='scales')
+
+        # Initialize guide params
+        for name, site in self.prototype_trace.iter_stochastic_nodes():
+            # Collect the shapes of unconstrained values.
+            # These may differ from the shapes of constrained values.
+            constrained_shape = site["value"].shape
+            unconstrained_shape = biject_to(site["fn"].support).inv(site["value"]).shape
+
+            event_dim = site["fn"].event_dim + len(unconstrained_shape) - len(constrained_shape)
+
+            site_batch_shape = site["fn"].batch_shape
+
+            self._unconstrained_shapes[name] = unconstrained_shape
+            unconstrained_shape_for_params = broadcast_shape(
+                unconstrained_shape,
+                site_batch_shape + (1,) * event_dim
+            )
+
+            # Collect independence contexts.
+            self._cond_indep_stacks[name] = site["cond_indep_stack"]
+
+            init_loc = site["value"].new_zeros(unconstrained_shape_for_params)
+            init_scale = site["value"].new_full(unconstrained_shape_for_params, self._init_scale)
+            _deep_setattr(self.locs, name, nn.Parameter(init_loc))
+            _deep_setattr(self.scales, name, PyroParam(init_scale, constraints.positive))
```

Comment:
On second thought I think it's easiest to support `init_loc_fn` in this PR. I think this should work:
```py
for name, site in self.prototype_trace.iter_stochastic_nodes():
    # Collect the shapes of unconstrained values.
    # These may differ from the shapes of constrained values.
    constrained_shape = site["value"].shape
    init_loc = biject_to(site["fn"].support).inv(site["value"].detach())
    self._unconstrained_shapes[name] = init_loc.shape

    # Collect independence contexts.
    self._cond_indep_stacks[name] = site["cond_indep_stack"]

    init_scale = torch.full_like(init_loc, self._init_scale)
    _deep_setattr(self.locs, name, nn.Parameter(init_loc))
    _deep_setattr(self.scales, name, PyroParam(init_scale, constraints.positive))
```
Then you can remove that `NotImplementedError` from `__init__` and I think even get rid of those `xfail_if_not_implemented`s in the tests.

Reply:
Awesome!  this works!  I removed the `xfail` 👍 .